### ISSUE TITLE
Let recursion

### ISSUE THREAD

**Issue by owickstrom** [oden-lang/oden#26]
https://oden-playground.herokuapp.com/p/67a451b1daa970b4e1bc7265532ff8e1

**Comment by marcomaggi** [comment]
A `let` syntax that allows recursion is called `letrec` in the Lisp world.  Have you chosen `let` on purpose because you want the compiler to detect recursion automatically?

More generally: do you want `let` forms that allow the programmer to explicitly select the region of visibility of variable syntactic bindings?

**Comment by owickstrom** [comment]
Yes, `let` should be able to handle recursive bindings as well, like `let ... in ...` in Haskell. There's no TCO  so it mostly affects the inferencer, and then the it uses the recursion in Go.

The Go output:

```go
var fact func (int) int
fact = func (n int) int {
	return (func() int {
		if n < 1 {
			return 1
		} else {
			return (n * fact((n - 1)))
		}
	})()
}
return fact(10)
```

Could you expand more on your second question? Not sure if I understand it correctly.

**Comment by marcomaggi** [comment]
> Could you expand more on your second question? Not sure if I understand it correctly.

It's like this: I am a schemer and I can testify that the `let` variants defined in the [revision 6 of the standard](http://www.r6rs.org/final/html/r6rs/r6rs-Z-H-14.html#node_sec_11.4.6) can be used effectively to specify the region of visibility of syntactic bindings.  I almost always find myself trying to use the binding syntax that specifies the stricter region of visibility.

Scheme's `let` does **not** allow two bindings with the same name; the order in which the right-hand side expressions is evaluated is unspecified (it is left to the compiler to decide); recursive bindings are forbidden.  So the following works:
```
(let ((x 1))
  (let ((x x))
   x)
=> 1
``` 

Scheme's `let*` does allow multiple bindings with the same name; the order in which the rhs expressions are evaluated is specified to be left-to-right; recursive bindings are forbidden.  We can define a `let*` syntax:
```
(let* ((?lhs1 ?rhs1)
        (?lhs2 ?rhs2))
  ?body)
```
with the recursive transformation:
```
(let ((?lhs1 ?rhs1))
  (let* ((?lhs2 ?rhs2))
    ?body))
```
that is: a `let*` syntax is transformed into nested `let` syntaxes.

Then there are `letrec` and `letrec*`; both forbid bindings with the same name; both allow recursive bindings; `letrec` does not specify the order in which the rhs expressions are evaluated, while `letrec*` specifies it as left-to-right.

An ugly but working implementation of `letrec` is with the following transformation:
```
(letrec ((?lhs1 ?rhs1)
            (?lhs2 ?rhs2))
  ?body)
==>
(let ((?lhs1 #<void>)
       (?lhs2 #<void>))
  ;;Does not specify ther order in which the ?rhs
  ;;are evaluated.
  (let ((tmp1 ?rhs1)
         (tmp2 ?rhs2))
    ?body))
```

An ugly but working implementation of `letrec*` is with the following transformation:
```
(letrec* ((?lhs1 ?rhs1)
              (?lhs2 ?rhs2))
  ?body)
==>
(let ((?lhs1 #<void>)
       (?lhs2 #<void>))
  ;;Does specify ther order in which the ?rhs
  ;;are evaluated.
  (set! ?lhs1 ?rhs1)
  (set! ?lhs2 ?rhs2)
  ?body)
```
the syntactic bindings do not need to be mutable if `set!` is an internal operation.

**Comment by marcomaggi** [comment]
Sorry, I realised that I have made a mistake; the `letrec` transformation is:
```
(letrec ((?lhs1 ?rhs1)
         (?lhs2 ?rhs2))
  ?body)
==>
(let ((?lhs1 #<void>)
      (?lhs2 #<void>))
  ;;Does not specify ther order in which the ?rhs
  ;;are evaluated.
  (let ((tmp1 ?rhs1)
        (tmp2 ?rhs2))
    (set! ?lhs1 tmp1)
    (set! ?lhs2 tmp2)) 
  ?body)
```

Also notice that once you have `letrec*` you also have internal definitions, because:
```
(define ?lhs1 ?rhs1)
(define ?lhs2 ?rhs2)
?expr
```
is equivalent to:
```
(letrec* ((?lhs1 ?rhs1)
          (?lhs2 ?rhs2))
  ?expr)
```
and it is easy to implement this transformation.

**Comment by owickstrom** [comment]
Thank you for the examples. I just had a brief understanding of the Scheme let-variations. Going to think this through, how it should work in Oden. It's a bigger decision than I first realized.

**Comment by Heimdell** [comment]
I recommend you to compile recursive functions into usages of `fixpoint` combinator (even if the language will not require to mark function as recursive).

so
```haskell
fact 0 = 1
fact n = n * fact (n - 1)
```

will become
```haskell
fact = fixpoint \self n -> case n of
    0 -> 1
    n -> n * self (n - 1)
```

There is a problem with mutual recursion, but it is solvable.

This will later allow you to clearly distinguish recursion and transform it into either for-loops with accum or unrolled loop or direct recursion in go. This also will help you to inline things.

**Comment by owickstrom** [comment]
@Heimdell Yeah, that sounds like a good idea. I have postponed these things for now, but it sounds like the right thing to do. Do you have any links to resources on doing the transformation from a fixpoint construct to an imperative programming language like Go?

**Comment by Heimdell** [comment]
I don't know any; but I know the devs of clojure-lang made a `loop`- `recur` construct which works as a fixpoint.
